server/src/main/java/org/elasticsearch/index/seqno/RetentionLeaseSyncAction.java

```diff
@@ -88,6 +96,53 @@ public RetentionLeaseSyncAction(
                 ThreadPool.Names.MANAGEMENT, false);
     }
 
+    @Override
+    protected void doExecute(Task parentTask, Request request, ActionListener<Response> listener) {
+        assert false : "use RetentionLeaseSyncAction#sync";
+    }
+
+    final void sync(ShardId shardId, String primaryAllocationId, long primaryTerm, RetentionLeases retentionLeases,
+                    ActionListener<ReplicationResponse> listener) {
+        final ThreadContext threadContext = threadPool.getThreadContext();
+        try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
+            threadContext.markAsSystemContext();
```

Comment:
It seems a bit weird to me to move to the system context here. Could we instead have these tasks run in the correct context from inception like this?

```diff
diff --git a/server/src/main/java/org/elasticsearch/common/util/concurrent/AbstractAsyncTask.java b/server/src/main/java/org/elasticsearch/common/util/concurrent/AbstractAsyncTask.java
index 1ef9a48..58e48f0 100644
--- a/server/src/main/java/org/elasticsearch/common/util/concurrent/AbstractAsyncTask.java
+++ b/server/src/main/java/org/elasticsearch/common/util/concurrent/AbstractAsyncTask.java
@@ -91,7 +91,7 @@ public abstract class AbstractAsyncTask implements Runnable, Closeable {
             if (logger.isTraceEnabled()) {
                 logger.trace("scheduling {} every {}", toString(), interval);
             }
-            cancellable = threadPool.schedule(this, interval, getThreadPool());
+            cancellable = threadPool.schedule(threadPool.preserveContext(this), interval, getThreadPool());
             isScheduledOrRunning = true;
         } else {
             logger.trace("scheduled {} disabled", toString());
```

Reply:
Thanks. Fixed in https://github.com/elastic/elasticsearch/pull/48430/commits/99ccfcbdc061b011f970364a0415b9783b8d70c7.